src/inputs/TreeSelectField/TreeSelectField.tsx

```diff
@@ -230,7 +267,33 @@ function TreeSelectFieldBase<O, V extends Value>(props: TreeSelectFieldProps<O,
       optionsLoading: false,
       allowCollapsing: true,
     };
-  });
+  }, [
+    initialOptions,
+    values,
+    chipDisplay,
+    getOptionLabel,
+    isReadOnly,
+    nothingSelectedText,
+    levelOptions,
+    getOptionValue,
+    getTopLevelSelections,
+  ]);
+
+  // Initialize the TreeFieldState
+  const [fieldState, setFieldState] = useState<TreeFieldState<O>>(() => initTreeFieldState());
+
+  // Reset the TreeFieldState if the values array changes and doesn't match the selectedOptions
+  useEffect(() => {
+    // if the values does not match the values in the fieldState, then update the fieldState
+    const selectedKeys = fieldState.selectedOptions.map((o) => valueToKey(getOptionValue(o)));
+    if (
+      values &&
+      (values.length !== selectedKeys.length || !values.every((v) => selectedKeys.includes(valueToKey(v))))
+    ) {
+      setFieldState(initTreeFieldState());
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Why is this disabled? Wouldn't we want to call the most-recent/most-correct `initTreeFieldState`, instead of the one with all of the captured-at-the-last-`values` render?

Kinda related, I generally think of `useCallback` being used to memoized functions that will respond to user input, i.e. "on click do something", and that "do something" needs to "capture over" some of our props.

It seems like most of the `useCallback`s here are not "react to user events (callbacks)" but just "regular helper methods", such that I wonder if they should just be helper methods at the bottom of the file, like instead of:

```
const fnThatUsesFoo = useCallback((a, b) => ..., [foo]);
```

We make a function at the bottom of the file:

```
function fnThatUsesFoo((foo, a, b) => ... {})
```

My suspicion is that would be cleaner, but I'm still kinda glancing/reading around.

Reply:
Yea, I think that's probably right. I can look to clean up.

Reply:
I disabled this initially because our only current use case of needing to reinitialize the state was when values change (that was the direct ask). We could do better. I'll see what else should trigger a re-initialization of the state.

Reply:
:+1: fwiw I generally believe/follow the logic of "only re-init when `values` changes" and having that be good enough, my concern was just that the lambda called a potentially-stale `initTreeFieldState`, such that the re-init wouldn't actually use the new `values` / other new props that might be relevant in its calculation/state creation.

So if this deps array is just `[values, initTreeFieldState]`, that's probably fine?...

Reply:
Yea, I get what you mean. I could potentially extract the logic for determining which values should be selected from the `initTreeFieldState` method and do that separately.
There's a decent amount of logic for determining the values to be selected for the use case where the caller only passes the leaf node ids. We need to check to see if the parent nodes should also be considered selected. Or if they only pass in the parent node ids, then we need to mark all children as selected.
This probably could use some more thought and refactoring to how/where we determine parent states. In this field we the parent selection state is only ever a reflection of its children's state (i.e. you can't have a parent selected but not its children).
I was hoping not to go too far into refactoring this component, but now you got me itching. 😄